packages/common/core-utils/src/assert.ts

```diff
@@ -51,9 +51,40 @@ export function assert(condition: boolean, message: string | number): asserts co
  * @internal
  */
 export function fail(message: string | number): never {
-	throw new Error(
+	const error = new Error(
 		typeof message === "number" ? `0x${message.toString(16).padStart(3, "0")}` : message,
 	);
+	onAssertionError(error);
+	throw error;
+}
+
+function onAssertionError(error: Error): void {
+	for (const handler of firstChanceAssertionHandler) {
+		handler(error);
+	}
+}
+
+const firstChanceAssertionHandler = new Set<(error: Error) => void>();
+
+/**
+ * Add a callback which can be used to report an assertion before it is thrown.
+ * @param handler - Called when an assertion occurs before the exception is thrown.
+ * @returns a function to remove the handler.
+ * @remarks
+ * This is mainly useful for triggering a `debugger;` statement and/or reporting telemetry to aid Fluid with diagnosing and fixing the bugs which cause asserts.
+ * It can also be used to record that an fatal error has occurred to help avoid complicating the diagnosis by doing more work after the error which might error again in a different way.
```

Comment:
```suggestion
 * It can also be used to record that a fatal error has occurred to help avoid complicating the diagnosis by doing more work after the error which might error again in a different way.
```

Reply:
This sentence is also a bit of a run-on and hard to parse.

Reply:
FWIW, the description you used in the changeset is easier to parse IMO